ibtk/src/solvers/impls/PoissonFACPreconditionerStrategy.cpp

```diff
@@ -360,6 +360,19 @@ PoissonFACPreconditionerStrategy::initializeOperatorState(const SAMRAIVectorReal
     d_coarsest_ln = solution.getCoarsestLevelNumber();
     d_finest_ln = solution.getFinestLevelNumber();
 
+#if !defined(NDEBUG)
+    // To prevent very obtuse errors later on, we check that the rhs index has the correct ghost cell width.
+    Pointer<PatchDescriptor<NDIM> > pd = VariableDatabase<NDIM>::getDatabase()->getPatchDescriptor();
+    const IntVector<NDIM>& gcw = pd->getPatchDataFactory(rhs_idx)->getGhostCellWidth();
+    if (gcw != d_gcw)
+    {
+        TBOX_ERROR(
+            d_object_name +
+                "::initializeOperatorState(): RHS index does not have the correct ghost width. RHS has ghost width of "
+            << gcw << ". Expecting a ghost cell width of " << d_gcw << ".\n");
+    }
+#endif
```

Comment:
What exactly are the errors this causes? I'm not that familiar with this preconditioner - do we end up applying it to the right-hand side at some point?

Can you document this requirement in the header?

Reply:
This just a multigrid preconditioner, so we need to restrict the RHS to coarser levels. The general issue is that `FACPreconditioner` calls the strategy implementations that coarsen or refine with copies or clones of the `b` vector, but `PoissonFACPreconditionerStrategy` caches the coarsen and refine schedules with the `d_scratch_idx`, which may have different ghost cell widths than whatever is inside of the `b` vector. In order to use the cached coarsen/refine schedules with different patch indices, the new patch indices need to be consistent with the cached patch indices. 

This is the simplest fix. The ghost cell width of `d_scratch_idx` is set in the constructor. If we require that the patch index in `b` has the same ghost width of `d_scratch_idx`, then all communication schedules will be consistent.

It's possible that we can relax the condition that `b` has the same ghost cell width as `d_scratch_idx`, but this assumption is asserted later in derived classes of `PoissonFACPreconditionerStrategy` (e.g. in `smoothError()` routines). This fix just pulls that assertion into the base class before any `solveSystem()` call occurs.

Reply:
I've added some documentation in the header about this requirement.

Reply:
I have also got burnt by these bugs more than once. It is good to throw error sooner than later.